src/CodeGen_LLVM.cpp

```diff
@@ -2457,6 +2459,150 @@ void CodeGen_LLVM::codegen_predicated_vector_load(const Load *op) {
     }
 }
 
+void CodeGen_LLVM::codegen_atomic_store(const Store *op) {
+    // TODO: predicated store
+    user_assert(is_one(op->predicate)) << "Atomic predicated store is not supported.\n";
+
+    // Detect whether we can describe this as an atomic-read-modify-write, 
+    // otherwise fallback to a compare-and-swap loop.
+    // Currently we only test for atomicAdd.
+    Expr val_expr = op->value;
+    Halide::Type value_type = op->value.type();
+    // For atomicAdd, we check if op->value - store[index] is independent of store.
+    // For llvm version < 9, the atomicRMW operations only support integers so we also check that.
+    Expr equiv_load = Load::make(value_type, op->name,
+                                 op->index,
+                                 Buffer<>(),
+                                 op->param,
+                                 op->predicate,
+                                 op->alignment);
+    Expr delta = simplify(common_subexpression_elimination(op->value - equiv_load));
+#if LLVM_VERSION >= 90
+    bool is_atomic_add = !expr_uses_var(delta, op->name);
+    if (is_ptx_atomic_store) {
```

Comment:
Rather than putting ptx-specific stuff in the base codegen class, how about a method on CodeGen_LLVM:

`virtual bool supports_atomic_add(const Type &t) const;`

Where the default implementation just checks if it's an Int, and Codegen_PTX_Dev overrides it to allow floats too.

Reply:
Done.